### PR TITLE
Script thread with no root document

### PR DESCRIPTION
**Pull request by asajeffrey**
This PR removes the single root document from the script thread, and replaces it by a lookup table from `PipelineId`s to `Document`s. This is needed if we're going to share script threads, as per #633.

The last commit is the one that matters, the ones before are #13646.

cc @jdm @Ms2ger @ConnorGBrewster 

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes do not require tests because refactoring

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14013)

### CODE REVIEW

File: components/script/script_thread.rs [hunk 1]
```diff
@@ -1218,24 +1256,22 @@ impl ScriptThread {
         let mut dom_tree_size = 0;
         let mut reports = vec![];
 
-        if let Some(root_context) = self.browsing_context.get() {
-            for it_context in root_context.iter() {
-                let current_url = it_context.active_document().url().to_string();
+        for document in self.documents.borrow().iter() {
+            let current_url = document.url().to_string();
```
Comment (frewsxcv) [inline components/script/script_thread.rs]: You might be able to get by with `let current_url = document.url().as_str();` here?

Comment (asajeffrey) [inline components/script/script_thread.rs]: It was a bit trickier than that, but I did manage to avoid double-copying the urls here.

**Comment by asajeffrey** [comment]
Rebased. @bors-servo try

**Comment by asajeffrey** [comment]
```
  ▶ CRASH [expected FAIL] /css-text-3_dev/html/line-break-strict-017.htm
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ called `Result::unwrap()` on an `Err` value: IoError(Error { repr: Os { code: 104, message: "Connection reset by peer" } }) (thread main, at ../src/libcore/result.rs:837)
  │ stack backtrace:
  │    0:     0x7ffc0078f3ad - backtrace::backtrace::trace::h16372ee7bf1517e5
  │    1:     0x7ffc0078fa42 - backtrace::capture::Backtrace::new::hcc43c50c4b11c693
  │    2:     0x7ffbff126e7d - servo::main::_{{closure}}::h993bd385ff269be0
  │    3:     0x7ffc00d52b74 - std::panicking::rust_panic_with_hook
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panicking.rs:452
  │    4:     0x7ffc00d52a34 - std::panicking::begin_panic<collections::string::String>
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panicking.rs:413
  │    5:     0x7ffc00d52959 - std::panicking::begin_panic_fmt
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panicking.rs:397
  │    6:     0x7ffc00d528e7 - std::panicking::rust_begin_panic
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panicking.rs:373
  │    7:     0x7ffc00d8f34d - core::panicking::panic_fmt
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libcore/panicking.rs:69
  │    8:     0x7ffc008202df - core::result::unwrap_failed::h97cf4d1c2597280e
  │    9:     0x7ffc008213d2 - webrender_traits::api::_<impl webrender_traits..RenderApiSender>::create_api::h8aae9b868a4fcac9
  │   10:     0x7ffbff1248fd - servo::main::hf0cef208a66e3646
  │   11:     0x7ffc00d5b1da - panic_unwind::__rust_maybe_catch_panic
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libpanic_unwind/lib.rs:97
  │   12:     0x7ffc00d51b9a - std::panicking::try<(),fn()>
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panicking.rs:332
  │                          - std::panic::catch_unwind<fn(),()>
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/panic.rs:351
  │                          - std::rt::lang_start
  │                         at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libstd/rt.rs:57
  │   13:     0x7ffbfca39f44 - __libc_start_main
  │   14:     0x7ffbff0f9909 - <unknown>
  │   15:                0x0 - <unknown>
  │ thread panicked while processing panic. aborting.
  └ thread panicked while processing panic. aborting.
```

**Comment by asajeffrey** [comment]
Looks like #13480.

**Comment by asajeffrey** [comment]
Rebased and squashed. @bors-servo try

**Comment by asajeffrey** [comment]
Rebased. #13646 has landed, so now this one can too.

File: components/script/script_thread.rs [hunk 2]
```diff
@@ -652,16 +711,12 @@ impl ScriptThread {
...
-        if let Some(context) = context {
-            for context in context.iter() {
-                // Only process a resize if layout is idle.
-                let window = context.active_window();
-                let resize_event = window.steal_resize_event();
-                match resize_event {
-                    Some(size) => resizes.push((window.upcast::<GlobalScope>().pipeline_id(), size)),
-                    None => ()
-                }
+        for document in self.documents.borrow().iter() {
```
Comment (jdm) [inline components/script/script_thread.rs]: If we can iterate over the keys and values simultaneously, we don't need to get the pipeline id from the window below.

File: components/script/script_thread.rs [hunk 3]
```diff
@@ -809,24 +864,20 @@ impl ScriptThread {
...
-                    // Reflow currently happens when explicitly invoked by code that
-                    // knows the document could have been modified. This should really
-                    // be driven by the compositor on an as-needed basis instead, to
-                    // minimize unnecessary work.
-                    window.reflow(ReflowGoal::ForDisplay,
-                                  ReflowQueryType::NoQuery,
-                                  ReflowReason::MissingExplicitReflow);
-                }
+        for document in self.documents.borrow().iter() {
+            let pending_reflows = document.window().get_pending_reflow_count();
```
Comment (jdm) [inline components/script/script_thread.rs]: Add a `let window = document.window()` here.

File: components/script/script_thread.rs [hunk 4]
```diff
@@ -1222,58 +1251,51 @@ impl ScriptThread {
...
-                    kind: ReportKind::ExplicitJemallocHeapSize,
-                    size: dom_tree_size,
-                });
-                urls.push(current_url);
+            for child in document.upcast::<Node>().traverse_preorder() {
+                dom_tree_size += heap_size_of_self_and_children(&*child);
             }
+            dom_tree_size += heap_size_of_self_and_children(document.window());
+
+            if reports.len() > 0 { path_seg.push_str(", "); }
```
Comment (jdm) [inline components/script/script_thread.rs]: nit: three lines for this, please.

File: components/script/script_thread.rs [hunk 5]
```diff
@@ -1659,20 +1644,15 @@ impl ScriptThread {
                                  incomplete.window_size);
         let frame_element = frame_element.r().map(Castable::upcast);
 
-        enum ContextToRemove {
-            Root,
-            Child(PipelineId),
-            None,
-        }
         struct AutoContextRemover<'a> {
```
Comment (jdm) [inline components/script/script_thread.rs]: We can remove this entirely; there are no early returns from this method any more.

File: components/script/script_thread.rs [hunk 6]
```diff
@@ -2259,35 +2186,27 @@ impl Drop for ScriptThread {
...
 
     // Drop our references to the JSContext and DOM objects.
-    for context in context_tree.iter() {
-        let window = context.active_window();
-        window.clear_js_runtime();
+    document.window().clear_js_runtime();
 
-        // Sever the connection between the global and the DOM tree
+    // If the window is for an active document, then
+    // sever the connection between the global and the DOM tree
```
Comment (jdm) [inline components/script/script_thread.rs]: The function is odd - `document.window()` can return a Window for a document that has no browsing context, but that doesn't mean that we want to shut down layout for that window. Perhaps this function should accept a `&Window` argument instead, and obtain the context from that.

File: components/script/webdriver_handlers.rs [hunk 7]
```diff
@@ -35,21 +33,19 @@ use msg::constellation_msg::PipelineId;
...
-        Some(context) => context,
-        None => return None
-    };
-
-    let document = context.active_document();
-    document.upcast::<Node>().traverse_preorder().find(|candidate| candidate.unique_id() == node_id)
+    documents.find_document(pipeline)
+        .and_then(|document| document.upcast::<Node>()
+                  .traverse_preorder()
+                  .find(|candidate| candidate.unique_id() == node_id))
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: nit: weird indentation.

File: components/script/webdriver_handlers.rs [hunk 8]
```diff
@@ -137,50 +131,41 @@ pub fn handle_get_frame_id(context: &BrowsingContext,
             }
         },
         WebDriverFrameId::Parent => {
-            let window = context.active_window();
-            Ok(window.parent())
+            documents.find_window(pipeline)
+                .map(|window| Ok(window.parent()))
+                .unwrap_or(Err(()))
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: `documents.find_window(pipeline).map(|window| window.parent()).ok_or(())`

File: components/script/webdriver_handlers.rs [hunk 9]
```diff
@@ -137,50 +131,41 @@ pub fn handle_get_frame_id(context: &BrowsingContext,
...
-pub fn handle_find_element_css(context: &BrowsingContext, _pipeline: PipelineId, selector: String,
+pub fn handle_find_element_css(documents: &Documents, pipeline: PipelineId, selector: String,
                                reply: IpcSender<Result<Option<String>, ()>>) {
-    reply.send(match context.active_document().QuerySelector(DOMString::from(selector)) {
-        Ok(node) => {
-            Ok(node.map(|x| x.upcast::<Node>().unique_id()))
-        }
-        Err(_) => Err(())
-    }).unwrap();
+    let node_id = documents.find_document(pipeline)
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]:
```rust
let node_id = documents.find_document(pipeline)
    .ok_or(())
    .and_then(|doc| doc.QuerySelector(DOMString::from(selector)).map_err(|_| ()))
    .map(|node| node.map(|x| x.upcast::<Node>().unique_id()));
```

File: components/script/webdriver_handlers.rs [hunk 10]
```diff
@@ -137,50 +131,41 @@ pub fn handle_get_frame_id(context: &BrowsingContext,
...
-            Ok(result)
-        },
-        Err(_) => {
-            Err(())
-        }
-    }).unwrap();
+    let node_ids = documents.find_document(pipeline)
+        .map(|doc| doc.QuerySelectorAll(DOMString::from(selector)).map_err(|_| ()))
+        .unwrap_or(Err(()))
+        .map(|nodes| nodes.iter().map(|x| x.upcast::<Node>().unique_id()).collect());
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: Same as previous.

File: components/script/webdriver_handlers.rs [hunk 11]
```diff
@@ -195,47 +180,62 @@ pub fn handle_focus_element(context: &BrowsingContext,
...
+    reply.send(documents.find_document(pipeline)
+               .and_then(|doc| doc.GetActiveElement())
+               .map(|elem| elem.upcast::<Node>().unique_id())).unwrap();
 }
 
-pub fn handle_get_cookies(context: &BrowsingContext,
-                         _pipeline: PipelineId,
+pub fn handle_get_cookies(documents: &Documents,
+                         pipeline: PipelineId,
                          reply: IpcSender<Vec<Serde<Cookie>>>) {
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: nit: indentation.

File: components/script_traits/script_msg.rs [hunk 12]
```diff
@@ -83,8 +83,9 @@ pub enum ScriptMsg {
     /// A new load has been requested, with an option to replace the current entry once loaded
     /// instead of adding a new entry.
     LoadUrl(PipelineId, LoadData, bool),
-    /// Dispatch a mozbrowser event to the parent of this pipeline.
-    /// The first PipelineId is for the parent, the second is for the originating pipeline.
+    /// Dispatch a mozbrowser event to a given iframe,
+    /// or to the window if no subpage id is provided.
```
Comment (jdm) [inline components/script_traits/script_msg.rs]: How would no subpage id be provided?

**Review by jdm — CHANGES_REQUESTED** [review]
Well done! There are a lot of straightforward changes here, and I like that some parts of the code actually became more straightforward with this change.

**Comment by asajeffrey** [comment]
@jdm thanks! I am trying to make the code easier to follow.

File: components/script/script_thread.rs (hunk 3 above)
Comment (asajeffrey) [inline components/script/script_thread.rs]: Done.

File: components/script/script_thread.rs (hunk 4 above)
Comment (asajeffrey) [inline components/script/script_thread.rs]: Done.

File: components/script/webdriver_handlers.rs (hunk 8 above)
Comment (asajeffrey) [inline components/script/webdriver_handlers.rs]: Done.

File: components/script/webdriver_handlers.rs (hunk 7 above)
Comment (asajeffrey) [inline components/script/webdriver_handlers.rs]: Hey, this is what emacs rust mode did! Reindented.

File: components/script/webdriver_handlers.rs (hunk 10 above)
Comment (asajeffrey) [inline components/script/webdriver_handlers.rs]: Done.

File: components/script_traits/script_msg.rs (hunk 12 above)
Comment (asajeffrey) [inline components/script_traits/script_msg.rs]: It happens, e.g., when we generate an error caused by a panic but we don't have a pipeline to blame it on, e.g. a panic from a resource thread.

Comment (jdm) [inline components/script_traits/script_msg.rs]: Ok, but the types involved are PipelineId and PipelineId, so I still don't understand what this means.

Comment (asajeffrey) [inline components/script_traits/script_msg.rs]: Oh oops, argh I reverted a code change but didn't revert the accompanying comment!

**Comment by jdm** [comment]
Squash and merge!

**Comment by asajeffrey** [comment]
Squashed. @bors-servo r=jdm

**Comment by asajeffrey** [comment]
Hmm, failing tests. Puts on debugging hat...

**Pull request merged**